### PR TITLE
242 add versionheader to drop message

### LINKED ISSUE #242: Implementation not compliant with Drop Message transport format spec
The [spec](https://github.com/Qabel/qabel-doc/wiki/Qabel-Client-Drop#transport-format) demands a version header which is currently missing in the implementation.

### PR DESCRIPTION
**Pull request by jasinai**
Concatenates the header with the encrypted message and removes it from the received message.
Fixes #242

### CODE REVIEW

**Comment by aBothe** [comment]
Looks fine to me :)

**Comment by jan-schreib** [comment]
What about a tests for your new methods? Trivial, but you know the drill. :)

**Comment by jasinai** [comment]
> What about a tests for your new methods? Trivial, but you know the drill. :)

Yes, perhaps it's a good idea

File: src/main/java/de/qabel/core/drop/DropController.java [hunk 1]
```diff
@@ -293,4 +298,30 @@ private String decryptDrop(byte[] cipher, QblPrimaryKeyPair keypair, QblSignPubl
 		CryptoUtils cu = new CryptoUtils();
 		return cu.decryptHybridAndValidateSignature(cipher, keypair, signkey);
 	}
+
+	/**
+	 * Concatenates the header byte array to the message byte array.
```
Comment (jan-schreib) [inline src/main/java/de/qabel/core/drop/DropController.java]: According to the signature, the header is not an array.

File: src/main/java/de/qabel/core/drop/DropController.java [hunk 2]
```diff
@@ -188,8 +191,9 @@ public DropResult send(DropMessage<? extends ModelObject> message, Collection<Co
 			cryptedMessage = encryptDrop(serialize(message),
 					contact.getEncryptionPublicKey(),
 					contact.getContactOwner().getPrimaryKeyPair().getSignKeyPairs().get(0));
+			byte[] cryptedMessageWithHeader = concatHeaderAndEncryptedMessage((byte) message.getVersion(), cryptedMessage);
```
Comment (cburkert) [inline src/main/java/de/qabel/core/drop/DropController.java]: The version returned by DropMessage's getVersion is not version required here.
We need to version of the **binary** message format here (see [doc](https://github.com/Qabel/qabel-doc/wiki/Qabel-Client-Drop#header)).
Please hard-code the version to 0 until we implement proper version multiplexing.

**Comment by jasinai** [comment]
@jan-schreib: do you really want these tests? I would test java internals. My tests look like this: https://gist.github.com/jasinai/f1ce89e74c640ab7a43f, so if you really want them, I have them.

**Comment by thechauffeur** [comment]
Later -- when we have different versions -- we will need a check with different handling according to the version and we will need tests for that.
For now consider writing one single test which creates a drop message, adds the header using your new method, removes the header using your other new method and then checks whether the drop message is the same message prior to adding the header. I would leave out the tests you mentioned above @jasinai.

After adding such a test I :+1: for the merge.

**Comment by jan-schreib** [comment]
@jasinai Sorry for the missunderstanding. Of course I don't want you to test Java internals. I don't want to make your work harder. :)
@thechauffeur is right about the test method and I :+1 for the merge also when finished.

**Comment by jasinai** [comment]
The test could be a little bit smaller, if serialize and deserialize from DropController were not private. :(

**Comment by jan-schreib** [comment]
Ok for now. @jasinai please create a ticket for the code duplication in the tests. It's not critical right now, but has to be done in the future.

**Pull request merged**